src/libraries/System.Net.Http/tests/FunctionalTests/Socks/SocksProxyTest.cs

```diff
@@ -0,0 +1,127 @@
+{
+    public abstract class SocksProxyTest : HttpClientHandlerTestBase
+    {
+        public SocksProxyTest(ITestOutputHelper helper) : base(helper) { }
+
+        private static string[] Hosts(string socksScheme) => socksScheme == "socks5"
+            ? new[] { "localhost", "127.0.0.1", "::1" }
+            : new[] { "localhost", "127.0.0.1" };
+
+        public static IEnumerable<object[]> TestLoopbackAsync_MemberData() =>
+            from scheme in new[] { "socks4", "socks4a", "socks5" }
+            from useSsl in BoolValues
+            from useAuth in BoolValues
+            from host in Hosts(scheme)
+            select new object[] { scheme, useSsl, useAuth, host };
+
+        [Theory]
+        [MemberData(nameof(TestLoopbackAsync_MemberData))]
+        public async Task TestLoopbackAsync(string scheme, bool useSsl, bool useAuth, string host)
+        {
+            if (useSsl && UseVersion == HttpVersion.Version20 && !PlatformDetection.SupportsAlpn)
+            {
+                return;
+            }
+
+            await LoopbackServerFactory.CreateClientAndServerAsync(
+                async uri =>
+                {
+                    using LoopbackSocksServer proxy = useAuth ? LoopbackSocksServer.Create("DOTNET", "424242") : LoopbackSocksServer.Create();
+                    using HttpClientHandler handler = CreateHttpClientHandler();
+                    using HttpClient client = CreateHttpClient(handler);
+
+                    client.DefaultVersionPolicy = HttpVersionPolicy.RequestVersionExact;
+
+                    handler.Proxy = new WebProxy($"{scheme}://localhost:{proxy.Port}");
+                    handler.ServerCertificateCustomValidationCallback = TestHelper.AllowAllCertificates;
+
+                    if (useAuth)
+                    {
+                        handler.Proxy.Credentials = new NetworkCredential("DOTNET", "424242");
+                    }
+
+                    uri = new UriBuilder(uri) { Host = host }.Uri;
+
+                    Assert.Equal("Echo", await client.GetStringAsync(uri));
+                },
+                async server => await server.HandleRequestAsync(content: "Echo"),
+                options: new GenericLoopbackOptions
+                {
+                    UseSsl = useSsl,
+                    Address = host == "::1" ? IPAddress.IPv6Loopback : IPAddress.Loopback
+                });
+        }
+
+        public static IEnumerable<object[]> TestExceptionalAsync_MemberData()
+        {
+            foreach (string scheme in new[] { "socks4", "socks4a" })
+                yield return new object[] { scheme, "::1", false, null };
+            foreach (string scheme in new[] { "socks4", "socks4a", "socks5" })
+            {
+                yield return new object[] { scheme, "localhost", true, null };
+                yield return new object[] { scheme, "localhost", true, new NetworkCredential("bad_username", "bad_password") };
+            }
+        }
```

Comment:
These tests should assert that correct exceptions are thrown for a given conditon - for example when a wrong password is used vs when a password isn't provided at all. The expected message could be added as part of MemberData here.